presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/AbstractResourceConfigurationManager.java

```diff
@@ -104,6 +110,39 @@ protected void validateRootGroups(ManagerSpec managerSpec)
         return selectors.build();
     }
 
+    @Override
+    public Optional<SelectionContext<VariableMap>> match(SelectionCriteria criteria)
+    {
+        Optional<SelectorResult<VariableMap>> selectorResult = pickSelectors(criteria);
+        if (selectorResult.isPresent() && selectorResult.get().getResourceGroupIdTemplate().isPresent()) {
+            ResourceGroupIdTemplate group = selectorResult.get().getResourceGroupIdTemplate().get();
+            return Optional.of(new SelectionContext<>(
+                    group.expandTemplate(selectorResult.get().getVariableMap()),
+                    selectorResult.get().getVariableMap(),
+                    group.getFirstDynamicSegment(),
+                    templateCreateQueryLimits(group)));
+        }
+        return Optional.empty();
+    }
+
+    protected synchronized void buildResourceGroupSpecsMap(Map<ResourceGroupIdTemplate, ResourceGroupSpec> resourceGroupSpecs, List<ResourceGroupSpec> childResourceGroups, Optional<ResourceGroupIdTemplate> parentId)
```

Comment:
+1 for the fact that instead of copying - you chose to reuse what we had across the classes - that said - we can make this better.

We should have a simple function which takes the `List<ResourceGroupSpec> rootGroups` as input and returns `Map<ResourceGroupIdTemplate, ResourceGroupSpec>` as the return type. Also, the previous code was incorrectly using `synchronized` there is no need for synchronized in this function.

Once we do this - the calling code (example in FileResourceGroupConfigurationManager) would change from 

```
Map<ResourceGroupIdTemplate, ResourceGroupSpec> resourceGroupSpecs = new HashMap<>();
buildResourceGroupSpecsMap(resourceGroupSpecs, rootGroups, Optional.empty());
this.resourceGroupSpecs = resourceGroupSpecs;
```

to

```
this.resourceGroupSpecs = buildResourceGroupSpecsMap(rootGroups);
``` 

Here is how I would do it - 

I am using an ImmutableMap - this helps us ensure that no one by mistakes changes the content of the `Map`

```
protected static Map<ResourceGroupIdTemplate, ResourceGroupSpec> buildResourceGroupSpecsMap(List<ResourceGroupSpec> rootGroups)
{
    ImmutableMap.Builder<ResourceGroupIdTemplate, ResourceGroupSpec> specBuilder = ImmutableMap.builder();
    buildResourceGroupSpecsMap(specBuilder, rootGroups, Optional.empty());
    return specBuilder.build();
}

private static void buildResourceGroupSpecsMap(ImmutableMap.Builder<ResourceGroupIdTemplate, ResourceGroupSpec> resourceGroupSpecs, List<ResourceGroupSpec> childResourceGroups, Optional<ResourceGroupIdTemplate> parentId)
{
    for (ResourceGroupSpec resourceGroupSpec : childResourceGroups) {
        ResourceGroupIdTemplate childId;
        if (parentId.isPresent()) {
            childId = ResourceGroupIdTemplate.forSubGroupNamed(parentId.get(), resourceGroupSpec.getName().toString());
        }   
        else {
            childId = new ResourceGroupIdTemplate(resourceGroupSpec.getName().toString());
        }   

        if (!resourceGroupSpec.getSubGroups().isEmpty()) {
            buildResourceGroupSpecsMap(resourceGroupSpecs, resourceGroupSpec.getSubGroups(), Optional.of(childId));
        }   
        resourceGroupSpecs.put(childId, resourceGroupSpec);
    }   
}
```